cloud_utils/vault.py

```diff
@@ -1,136 +1,51 @@
+            secrets_data = {}
+        secrets_data.update(secrets)
+        client.secrets.kv.v2.create_or_update_secret(
+            path=path,
+            secret=secrets_data,
+            mount_point=_MOUNT_POINT,
         )
 
-    return _pod_identity_token
+    return write_or_update_secrets
 
 
-def _make_vault_headers(
-    token: Callable[[], Coroutine[Any, Any, str]],
-) -> Callable[[], Coroutine[Any, Any, Dict[str, str]]]:
-    async def _vault_headers() -> Dict[str, str]:
-        return {
-            "X-Vault-Token": await token(),
-        }
+def make_vault(host: str, token: Optional[str], role: Optional[str]):
+    if not token and not role:
+        raise Exception("Must specify either token or role")
 
-    return _vault_headers
-
-
-def _make_read_key(
-    base_vault_url: str,
-    headers: Callable[[], Coroutine[Any, Any, dict]],
-):
-    async def read_key(path: str) -> Dict[str, str]:
-        return gamla.pipe(
-            await gamla.get_async_with_headers(
-                (await headers()),
-                30,
-                f"{base_vault_url}/secret/data/{path}",
-            ),
-            lambda response: response.json(),
-            gamla.excepts(
-                KeyError,
-                gamla.make_raise(KeyDoesNotExist),
-                gamla.get_in(["data", "data"]),
-            ),
+    if token:
+        client = hvac.Client(url=host, token=token)
+    else:
+        client = hvac.Client(url=host)
+        Kubernetes(client.adapter).login(
+            role=role,
+            jwt=open("/var/run/secrets/kubernetes.io/serviceaccount/token").read(),
         )
 
-    return read_key
-
-
-def _make_write_key(
-    base_vault_url: str,
-    headers: Callable[[], Coroutine[Any, Any, dict]],
-):
-    async def write_key(path: str, value: Dict[str, str]):
-        return await gamla.post_json_with_extra_headers_async(
-            (await headers()),
-            30,
-            f"{base_vault_url}/secret/data/{path}",
-            {"data": value},
-        )
-
-    return write_key
-
-
-def _make_update_key(read_key: Callable, write_key: Callable):
-    async def update_key(path: str, value: Dict[str, str]):
-        current_value = {}
-        try:
-            current_value = await read_key(path)
-        except KeyDoesNotExist:
-            pass
-        return await write_key(path, gamla.merge(current_value, value))
-
-    return update_key
-
-
-def make_vault(host: str, role: str, token: Optional[str]):
-    async def get_token():
-        return token
-
-    base_vault_url = f"{host}/v1"
-    headers = _make_vault_headers(
-        get_token if token else _make_pod_identity_token(role, base_vault_url),
-    )
-
-    write_key, read_key = _make_write_key(base_vault_url, headers), _make_read_key(
-        base_vault_url,
-        headers,
-    )
-    return write_key, read_key, _make_update_key(read_key, write_key)
+    read_all_secrets = _build_read_all_secrets(client)
```

Comment:
Name is misleading here. This only read a single secret where secret is a dictionary of key, value pairs.
A better name is "read_secret"